src/Model/Attributes.php

```diff
@@ -178,13 +166,23 @@ public function toArray()
      *
      * @return mixed
      */
-    public function __get($key)
+    public function &__get($key)
     {
+        if ('attributes' === $key) {
+            return $this->attributes;
+        }
+
         if ($this->mutable && $this->hasMutatorMethod($key, 'get')) {
-            return $this->{$this->buildMutatorMethod($key, 'get')}();
+            $this->mutableCache[$key] = $this->{$this->buildMutatorMethod($key, 'get')}();
```

Comment:
when we try to do this:
```php
$model->array['key'] = 'value';
```
it fails with an ugly `PHP Notice:  Indirect modification of overloaded property Model::$array has no effect`.
The reason is that `__set()` is never called on these cases. what php does is something like it:
```php
$array = $model->array; // calling __get('array')
$array['key'] = 'value';
```
to "fix" these cases, we use `&__get` (as reference), so php knows where to store the value, and it "magically" works.
[(see this)](https://stackoverflow.com/questions/45314024/php-indirect-modification-of-overloaded-property-with-arrays)

but as a side effect, all returns of method `&__get` must be an attribute/variable, so that's why
we are using cache for these cases, and also setting `null` for attributes with no value (the other question).

since we are not saving `nulls` to database anymore, there is no problem if something is set to `null` but never used :)
(that's also why there is a `$user->unknown` check on integration test, to make sure that this is not going to be persisted)

Reply:
thanks for the explanation